Mix packages and modules in command line parameters for Pylint

Pylint is meant to be called from the command line. The usage is
```
pylint [options] modules_or_packages
```
You shoud passed modules OR packages. If you pass mix of modules and packages you should check that its not intercect between each other, because Pylint searches modules in packages and we got double messages in report (can live with it) or code duplication message in bad case (its a problem).

Test project (https://drive.google.com/file/d/13n1RZFdgwitfgUDOv76ZMYOPIrBYVsAT/view?usp=sharing):
```
a/__init__.py
a/module.py
a/b/__init__.py
a/b/module.py
b/__init__.py
b/module.py
```

Run pyb analyze -X:
```
[INFO]  Executing pylint on project sources
[DEBUG] Executing command: 'c:\\work\\projects\\misc\\pybuilder-pylint-issue\\.pybuilder\\plugins\\cpython-3.9.7.final.0\\scripts\\pylint.EXE' '--max-line-length=100' '--no-docstring-rgx=.*' 'a.module' 'a' 'a.b.module' 'a.b' 'b.module' 'b'
```

Report:
```
************* Module a.module
c:\work\projects\misc\pybuilder-pylint-issue\src\main\python\a\module.py:1:0: C0114: Missing module docstring (missing-module-docstring)
c:\work\projects\misc\pybuilder-pylint-issue\src\main\python\a\module.py:1:0: W0611: Unused import sys (unused-import)
c:\work\projects\misc\pybuilder-pylint-issue\src\main\python\a\module.py:1:0: C0114: Missing module docstring (missing-module-docstring)
c:\work\projects\misc\pybuilder-pylint-issue\src\main\python\a\module.py:1:0: W0611: Unused import sys (unused-import)
************* Module a.b.module
c:\work\projects\misc\pybuilder-pylint-issue\src\main\python\a\b\module.py:1:0: C0114: Missing module docstring (missing-module-docstring)
c:\work\projects\misc\pybuilder-pylint-issue\src\main\python\a\b\module.py:1:0: W0611: Unused import sys (unused-import)
c:\work\projects\misc\pybuilder-pylint-issue\src\main\python\a\b\module.py:1:0: C0114: Missing module docstring (missing-module-docstring)
c:\work\projects\misc\pybuilder-pylint-issue\src\main\python\a\b\module.py:1:0: W0611: Unused import sys (unused-import)
c:\work\projects\misc\pybuilder-pylint-issue\src\main\python\a\b\module.py:1:0: C0114: Missing module docstring (missing-module-docstring)
c:\work\projects\misc\pybuilder-pylint-issue\src\main\python\a\b\module.py:1:0: W0611: Unused import sys (unused-import)
************* Module b.module
c:\work\projects\misc\pybuilder-pylint-issue\src\main\python\b\module.py:1:0: C0114: Missing module docstring (missing-module-docstring)
c:\work\projects\misc\pybuilder-pylint-issue\src\main\python\b\module.py:5:0: C0103: Function name "b" doesn't conform to snake_case naming style (invalid-name)
c:\work\projects\misc\pybuilder-pylint-issue\src\main\python\b\module.py:1:0: W0611: Unused import sys (unused-import)
c:\work\projects\misc\pybuilder-pylint-issue\src\main\python\b\module.py:2:0: W0611: Unused module imported from a as a (unused-import)
c:\work\projects\misc\pybuilder-pylint-issue\src\main\python\b\module.py:1:0: C0114: Missing module docstring (missing-module-docstring)
c:\work\projects\misc\pybuilder-pylint-issue\src\main\python\b\module.py:5:0: C0103: Function name "b" doesn't conform to snake_case naming style (invalid-name)
c:\work\projects\misc\pybuilder-pylint-issue\src\main\python\b\module.py:1:0: W0611: Unused import sys (unused-import)
c:\work\projects\misc\pybuilder-pylint-issue\src\main\python\b\module.py:2:0: W0611: Unused module imported from a as a (unused-import)
c:\work\projects\misc\pybuilder-pylint-issue\src\main\python\b\module.py:1:0: R0801: Similar lines in 2 files
==b.module:[0:6]
==b.module:[0:6]
import sys
from a import module as a
from a.b import module as ab

def b():
    print('b') (duplicate-code)

------------------------------------------------------------------

Your code has been rated at 0.50/10 (previous run: 0.50/10, +0.00)
```

If I execute Pylint and pass only modules or only packages - all works great.

`pylint 'a' 'b'`:

```
************* Module a.module
a\module.py:1:0: C0114: Missing module docstring (missing-module-docstring)
a\module.py:1:0: W0611: Unused import sys (unused-import)
************* Module a.b.module
a\b\module.py:1:0: C0114: Missing module docstring (missing-module-docstring)
a\b\module.py:1:0: W0611: Unused import sys (unused-import)
************* Module b.module
b\module.py:1:0: C0114: Missing module docstring (missing-module-docstring)
b\module.py:5:0: C0103: Function name "b" doesn't conform to snake_case naming style (invalid-name)
b\module.py:5:0: C0116: Missing function or method docstring (missing-function-docstring)
b\module.py:1:0: W0611: Unused import sys (unused-import)
b\module.py:2:0: W0611: Unused module imported from a as a (unused-import)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```

`pylint 'a.module' 'a.b.module' 'b.module'`:

```
************* Module a.module
a\module.py:1:0: C0114: Missing module docstring (missing-module-docstring)
a\module.py:1:0: W0611: Unused import sys (unused-import)
************* Module a.b.module
a\b\module.py:1:0: C0114: Missing module docstring (missing-module-docstring)
a\b\module.py:1:0: W0611: Unused import sys (unused-import)
************* Module b.module
b\module.py:1:0: C0114: Missing module docstring (missing-module-docstring)
b\module.py:5:0: C0103: Function name "b" doesn't conform to snake_case naming style (invalid-name)
b\module.py:5:0: C0116: Missing function or method docstring (missing-function-docstring)
b\module.py:1:0: W0611: Unused import sys (unused-import)
b\module.py:2:0: W0611: Unused module imported from a as a (unused-import)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```

--- a/src/integrationtest/python/issue_818_tests.py
+++ b/src/integrationtest/python/issue_818_tests.py
@@ -1,0 +1,58 @@
+#   -*- coding: utf-8 -*-
+#
+#   This file is part of PyBuilder
+#
+#   Copyright 2011-2021 PyBuilder Team
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+import textwrap
+import unittest
+
+from itest_support import IntegrationTestSupport
+from pybuilder.errors import BuildFailedException
+
+
+class Issue818Test(IntegrationTestSupport):
+    def test(self):
+        self.write_build_file("""
+from pybuilder.core import use_plugin, init
+
+use_plugin("python.core")
+use_plugin("python.pylint")
+
+@init
+def init (project):
+    project.set_property("verbose", True)
+    project.set_property("pylint_break_build", True)
+        """)
+
+        self.create_directory("src/main/python/a")
+        self.write_file("src/main/python/a/__init__.py", textwrap.dedent(
+            """
+            print("Hello from the package")
+            """))
+        self.write_file("src/main/python/a/module.py", textwrap.dedent(
+            """
+            print("Hello from the module")
+            """))
+        reactor = self.prepare_reactor()
+
+        with self.assertRaises(BuildFailedException) as raised_e:
+            reactor.build("analyze")
+
+        self.assertEqual(raised_e.exception.message, "Pylint found 2 warning(s).")
+
+
+if __name__ == "__main__":
+    unittest.main()

--- a/src/main/python/pybuilder/plugins/python/pylint_plugin.py
+++ b/src/main/python/pybuilder/plugins/python/pylint_plugin.py
@@ -47,17 +47,24 @@ def execute_pylint(project, logger, reactor):
     logger.info("Executing pylint on project sources")
 
     command_and_arguments = ["pylint"] + project.get_property("pylint_options")
-    result_tuple = execute_tool_on_modules(project, "pylint", reactor.pybuilder_venv, command_and_arguments, True)
+    result_tuple = execute_tool_on_modules(project, "pylint", reactor.pybuilder_venv, command_and_arguments,
+                                           extend_pythonpath=True,
+                                           include_package_modules=False)
 
-    if project.get_property("pylint_break_build"):
-        report_file = result_tuple[1]  # access by position to avoid changing mocking behaviour
+    report_file = result_tuple[1]  # access by position to avoid changing mocking behaviour
 
-        warnings = [line
-                    for line in read_file(report_file)
-                    if line.find('.py:') >= 0]
-        warning_count = len(warnings)
+    warnings = [line.rstrip()
+                for line in read_file(report_file)
+                if line.find(".py:") >= 0]
+    warning_count = len(warnings)
 
-        if warning_count > 0:
-            message = "Pylint found {} warning(s).".format(warning_count)
+    if warning_count:
+        for warning in warnings:
+            logger.warn("PyLint: %s", warning)
+
+        message = "Pylint found {} warning(s).".format(warning_count)
+        if project.get_property("pylint_break_build"):
             logger.error(message)
             raise BuildFailedException(message)
+        else:
+            logger.warn(message)

--- a/src/main/python/pybuilder/plugins/python/python_plugin_helper.py
+++ b/src/main/python/pybuilder/plugins/python/python_plugin_helper.py
@@ -92,9 +92,16 @@ def execute_tool_on_source_files(project, name, python_env, command_and_argument
     return execution_result
 
 
-def execute_tool_on_modules(project, name, python_env, command_and_arguments, extend_pythonpath=True):
+def execute_tool_on_modules(project, name, python_env, command_and_arguments,
+                            extend_pythonpath=True,
+                            include_packages=True,
+                            include_package_modules=True,
+                            include_namespace_modules=True):
     source_dir = project.expand_path("$dir_source_main_python")
-    modules = discover_modules(source_dir)
+    modules = discover_modules(source_dir,
+                               include_packages=include_packages,
+                               include_package_modules=include_package_modules,
+                               include_namespace_modules=include_namespace_modules)
     command = as_list(command_and_arguments) + modules
 
     report_file = project.expand_path("$dir_reports/%s" % name)

--- a/src/main/python/pybuilder/utils.py
+++ b/src/main/python/pybuilder/utils.py
@@ -111,7 +111,7 @@ def remove_leading_slash_or_dot_from_path(path):
 
 def remove_python_source_suffix(file_name):
     if file_name.endswith(".py"):
-        return file_name[0:-len(".py")]
+        return file_name[0:-3]
     return file_name
 
 
@@ -130,22 +130,51 @@ def discover_module_files_matching(source_path, module_glob):
     return result
 
 
-def discover_modules(source_path, suffix=".py"):
-    return discover_modules_matching(source_path, "*{0}".format(suffix))
+def discover_modules(source_path, suffix=".py",
+                     include_packages=True,
+                     include_package_modules=True,
+                     include_namespace_modules=True
+                     ):
+    return discover_modules_matching(source_path, "*{0}".format(suffix),
+                                     include_packages=include_packages,
+                                     include_package_modules=include_package_modules,
+                                     include_namespace_modules=include_namespace_modules)
 
 
-def discover_modules_matching(source_path, module_glob):
+def discover_modules_matching(source_path, module_glob,
+                              include_packages=True,
+                              include_package_modules=True,
+                              include_namespace_modules=True):
     result = []
     if not module_glob.endswith(".py"):
         module_glob += ".py"
-    for module_file_path in discover_files_matching(source_path, module_glob):
-        relative_module_file_path = os.path.relpath(module_file_path, source_path)
-        relative_module_file_path = relative_module_file_path.replace(os.sep, ".")
-        module_file = remove_leading_slash_or_dot_from_path(relative_module_file_path)
-        module_name = remove_python_source_suffix(module_file)
-        if module_name.endswith(".__init__"):
-            module_name = module_name.replace(".__init__", "")
-        result.append(module_name)
+
+    for root, dirs, files in os.walk(source_path, followlinks=True):
+        is_package = False
+        for file_name in files:
+            if file_name == "__init__.py":
+                is_package = True
+                break
+
+        for file_name in files:
+            if fnmatch.fnmatch(file_name, module_glob):
+                module_file_path = jp(root, file_name)
+                relative_module_file_path = os.path.relpath(module_file_path, source_path)
+                relative_module_file_path = relative_module_file_path.replace(os.sep, ".")
+                module_file = remove_leading_slash_or_dot_from_path(relative_module_file_path)
+                module_name = remove_python_source_suffix(module_file)
+
+                add_module = False
+                if file_name == "__init__.py":
+                    if include_packages:
+                        module_name = module_name[:-9]  # len(".__init__")
+                        add_module = True
+                else:
+                    add_module = (not is_package and include_namespace_modules or
+                                  is_package and include_package_modules)
+
+                if add_module:
+                    result.append(module_name)
     return result
 
 
@@ -154,7 +183,7 @@ def discover_files(start_dir, suffix):
 
 
 def discover_files_matching(start_dir, file_glob, exclude_file_glob=None):
-    for root, _, files in os.walk(start_dir):
+    for root, _, files in os.walk(start_dir, followlinks=True):
         for file_name in files:
             if exclude_file_glob:
                 for exclude_pat in as_list(exclude_file_glob):

--- a/src/unittest/python/plugins/python/pylint_plugin_tests.py
+++ b/src/unittest/python/plugins/python/pylint_plugin_tests.py
@@ -59,7 +59,8 @@ class PylintPluginTests(TestCase):
 
         self.reactor.pybuilder_venv.verify_can_execute.assert_called_with(['pylint'], 'pylint', 'plugin python.pylint')
 
-    def test_should_run_pylint_with_default_options(self):
+    @patch('pybuilder.plugins.python.pylint_plugin.read_file', return_value=PYLINT_NORMAL_OUTPUT)
+    def test_should_run_pylint_with_default_options(self, *_):
         init_pylint(self.project)
 
         execute_pylint(self.project, Mock(Logger), self.reactor)
@@ -67,7 +68,8 @@ class PylintPluginTests(TestCase):
         self.reactor.pybuilder_venv.execute_command.assert_called_with(["pylint"] + DEFAULT_PYLINT_OPTIONS, ANY,
                                                                        env=ANY)
 
-    def test_should_run_pylint_with_custom_options(self):
+    @patch('pybuilder.plugins.python.pylint_plugin.read_file', return_value=PYLINT_NORMAL_OUTPUT)
+    def test_should_run_pylint_with_custom_options(self, *_):
         init_pylint(self.project)
         self.project.set_property("pylint_options", ["--test", "-f", "--x=y"])
 

--- a/src/unittest/python/utils_tests.py
+++ b/src/unittest/python/utils_tests.py
@@ -165,26 +165,26 @@ class DiscoverFilesTest(unittest.TestCase):
         expected_result = [nc("spam/spam.py"), nc("spam/eggs.py")]
         actual_result = set(discover_files("spam", ".py"))
         self.assertEqual(set(expected_result), actual_result)
-        walk.assert_called_with("spam")
+        walk.assert_called_with("spam", followlinks=True)
 
     @patch("pybuilder.utils.os.walk", return_value=[("spam", [], fake_dir_contents)])
     def test_should_only_return_py_glob(self, walk):
         expected_result = [nc("spam/readme.md")]
         actual_result = set(discover_files_matching("spam", "readme.?d"))
         self.assertEqual(set(expected_result), actual_result)
-        walk.assert_called_with("spam")
+        walk.assert_called_with("spam", followlinks=True)
 
 
 class DiscoverModulesTest(unittest.TestCase):
     @patch("pybuilder.utils.os.walk", return_value=[("spam", [], ["eggs.pi"])])
     def test_should_return_empty_list_when_directory_contains_single_file_not_matching_suffix(self, walk):
         self.assertEqual([], discover_modules("spam", ".py"))
-        walk.assert_called_with("spam")
+        walk.assert_called_with("spam", followlinks=True)
 
     @patch("pybuilder.utils.os.walk", return_value=[("spam", [], ["eggs.py"])])
     def test_should_return_list_with_single_module_when_directory_contains_single_file(self, walk):
         self.assertEqual(["eggs"], discover_modules("spam", ".py"))
-        walk.assert_called_with("spam")
+        walk.assert_called_with("spam", followlinks=True)
 
     @patch("pybuilder.utils.os.walk", return_value=[("pet_shop", [],
                                                      ["parrot.txt", "parrot.py", "parrot.pyc", "parrot.py~",
@@ -193,27 +193,63 @@ class DiscoverModulesTest(unittest.TestCase):
         expected_result = ["parrot"]
         actual_result = discover_modules_matching("pet_shop", "*parrot*")
         self.assertEqual(set(expected_result), set(actual_result))
-        walk.assert_called_with("pet_shop")
+        walk.assert_called_with("pet_shop", followlinks=True)
 
     @patch("pybuilder.utils.os.walk", return_value=[("spam", [], ["eggs.py"])])
     def test_glob_should_return_list_with_single_module_when_directory_contains_single_file(self, walk):
         self.assertEqual(["eggs"], discover_modules_matching("spam", "*"))
-        walk.assert_called_with("spam")
+        walk.assert_called_with("spam", followlinks=True)
 
     @patch("pybuilder.utils.os.walk", return_value=[("spam", ["eggs"], []),
                                                     ("spam/eggs", [], ["__init__.py"])])
     def test_glob_should_return_list_with_single_module_when_directory_contains_package(self, walk):
         self.assertEqual(["eggs"], discover_modules_matching("spam", "*"))
 
-        walk.assert_called_with("spam")
+        walk.assert_called_with("spam", followlinks=True)
 
-    @patch("pybuilder.utils.discover_files_matching", return_value=['/path/to/tests/reactor_tests.py'])
+    @patch("pybuilder.utils.os.walk", return_value=[("/path/to/tests", [], ["reactor_tests.py"])])
     def test_should_not_eat_first_character_of_modules_when_source_path_ends_with_slash(self, _):
         self.assertEqual(["reactor_tests"], discover_modules_matching("/path/to/tests/", "*"))
 
-    @patch("pybuilder.utils.discover_files_matching", return_value=['/path/to/tests/reactor_tests.py'])
+    @patch("pybuilder.utils.os.walk", return_value=[("/path/to/tests", [], ["reactor_tests.py"])])
     def test_should_honor_suffix_without_stripping_it_from_module_names(self, _):
         self.assertEqual(["reactor_tests"], discover_modules_matching("/path/to/tests/", "*_tests"))
+
+    @patch("pybuilder.utils.os.walk",
+           return_value=[("python", ["a", "b", "name"], []),
+                         ("python/a", ["b"], ["module.py", "__init__.py"]),
+                         ("python/a/b", [], ["module.py", "__init__.py"]),
+                         ("python/b", [], ["module.py", "__init__.py"]),
+                         ("python/name", ["space"], []),
+                         ("python/name/space", ["x"], ["module.py"]),
+                         ("python/name/space/x", [], ["__init__.py", "module.py"])])
+    def test_packages_and_ns_modules_only(self, _):
+        self.assertEqual(["a", "a.b", "b", "name.space.module", "name.space.x"],
+                         discover_modules_matching("python", "*", include_package_modules=False))
+
+    @patch("pybuilder.utils.os.walk",
+           return_value=[("python", ["a", "b", "name"], []),
+                         ("python/a", ["b"], ["module.py", "__init__.py"]),
+                         ("python/a/b", [], ["module.py", "__init__.py"]),
+                         ("python/b", [], ["module.py", "__init__.py"]),
+                         ("python/name", ["space"], []),
+                         ("python/name/space", ["x"], ["module.py"]),
+                         ("python/name/space/x", [], ["__init__.py", "module.py"])])
+    def test_non_package_modules_and_ns_modules_only(self, _):
+        self.assertEqual(["a.module", "a.b.module", "b.module", "name.space.module", "name.space.x.module"],
+                         discover_modules_matching("python", "*", include_packages=False))
+
+    @patch("pybuilder.utils.os.walk",
+           return_value=[("python", ["a", "b", "name"], []),
+                         ("python/a", ["b"], ["module.py", "__init__.py"]),
+                         ("python/a/b", [], ["module.py", "__init__.py"]),
+                         ("python/b", [], ["module.py", "__init__.py"]),
+                         ("python/name", ["space"], []),
+                         ("python/name/space", ["x"], ["module.py"]),
+                         ("python/name/space/x", [], ["__init__.py", "module.py"])])
+    def test_packages_and_package_modules_only(self, _):
+        self.assertEqual(["a.module", "a", "a.b.module", "a.b", "b.module", "b", "name.space.x", "name.space.x.module"],
+                         discover_modules_matching("python", "*", include_namespace_modules=False))
 
 
 class ApplyOnFilesTest(unittest.TestCase):

